main/src/hadd.cxx

```diff
@@ -90,281 +98,382 @@
+   std::optional<TString> fCacheSize;
+   std::optional<ROOT::TIOFeatures> fFeatures;
+   std::optional<IntFlag_t> fMaxOpenedFiles;
+   std::optional<IntFlag_t> fVerbosity;
+   std::optional<IntFlag_t> fCompressionSettings;
+
+   int fOutputArgIdx;
+   int fFirstInputIdx;
+};
+
+enum class EFlagResult { kIgnored, kParsed, kError };
+
+static EFlagResult FlagToggle(const char *arg, const char *flagStr, bool &flagOut)
 {
-   if ( argc < 3 || "-h" == std::string(argv[1]) || "--help" == std::string(argv[1]) ) {
-         fprintf(stderr, kCommandLineOptionsHelp);
-         return (argc == 2 && ("-h" == std::string(argv[1]) || "--help" == std::string(argv[1]))) ? 0 : 1;
+   const auto argLen = strlen(arg);
+   const auto flagLen = strlen(flagStr);
+   if (argLen == flagLen && strncmp(arg, flagStr, flagLen) == 0) {
+      if (flagOut)
+         std::cerr << "[warn] Duplicate flag: " << flagStr << "\n";
+      flagOut = true;
    }
+   return flagOut ? EFlagResult::kParsed : EFlagResult::kIgnored;
+}
+
+// NOTE: not using std::stoi or similar because they have bad error checking.
+// std::stoi will happily parse "120notvalid" as 120.
+static std::optional<IntFlag_t> StrToUInt(const char *str)
+{
+   if (!str)
+      return {};
+
+   uint32_t res = 0;
+   do {
+      if (!isdigit(*str))
+         return {};
+      if (res * 10 < res) // overflow is an error
+         return {};
+      res *= 10;
+      res += *str - '0';
+   } while (*++str);
+
+   return res;
+}
+
+template <typename T>
+static std::optional<T> ConvertArg(const char *);
+
+template <>
+std::optional<std::string> ConvertArg<std::string>(const char *arg)
+{
+   return arg;
+}
 
+template <>
+std::optional<IntFlag_t> ConvertArg<IntFlag_t>(const char *arg)
+{
+   return StrToUInt(arg);
+}
+
+template <>
+std::optional<ROOT::TIOFeatures> ConvertArg<ROOT::TIOFeatures>(const char *arg)
+{
    ROOT::TIOFeatures features;
-   Bool_t append = kFALSE;
-   Bool_t force = kFALSE;
-   Bool_t skip_errors = kFALSE;
-   Bool_t reoptimize = kFALSE;
-   Bool_t noTrees = kFALSE;
-   Bool_t keepCompressionAsIs = kFALSE;
-   Bool_t useFirstInputCompression = kFALSE;
-   Bool_t multiproc = kFALSE;
-   Bool_t debug = kFALSE;
-   Int_t maxopenedfiles = 0;
-   Int_t verbosity = 99;
+   std::stringstream ss;
+   ss.str(arg);
+   std::string item;
+   while (std::getline(ss, item, ',')) {
+      if (!features.Set(item))
+         std::cerr << "Ignoring unknown feature request: " << item << std::endl;
+   }
+   return features;
+}
+
+static std::optional<IntFlag_t> ConvertNProcesses(const char *arg)
+{
+   auto np = ConvertArg<IntFlag_t>(arg);
+   if (!np) {
+      // By returning a non-nullopt, we enable multiprocessing.
+      // Since 0 is not a valid number of processes, hadd will default to the number of cpus.
+      np = 0;
+   }
+   if (*np == 0) {
+      std::cerr << "Error: the number of parallel processes passed after -j is invalid: " << arg
+                << ". We will use the system maximum.\n";
+   }
```

Comment:
If it's invalid, why we default to the system maximum instead of aborting and asking the user to provide a valid number?

Reply:
I didn't want to change the existing logic too much lest we break some user scripts. But if that's an acceptable risk I'm more than happy to change it.

Reply:
Thanks for the clarification, I wasn't able to infer that it was a pre-existing behaviour. In this case then it might be wiser to just keep it as is (the other comment about formatting of error/warning messages still stands)